sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt

```diff
@@ -420,9 +422,12 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
     if (renderThreadPrepared && !paused) {
       draw()
     }
-    // we drain queue despite buffers swapped or not
-    drainQueue(nonRenderEventQueue)
     awaitingNextVsync = false
+    // It's critical to drain queue after setting awaitingNextVsync to `false` as some scheduled tasks
+    // may recursively schedule other tasks in the same callchain. If awaitingNextVsync is `true` -
+    // this leads to concurrent queue modification that may lead to deadlock in core. Instead we will
+    // always schedule recursive tasks on the render thread making sure they are not using the same callchain.
```

Comment:
It's critical to drain queue after setting awaitingNextVsync to `false` as some tasks may recursively schedule other tasks when executed. 
With `awaitingNextVsync = false` we will always schedule recursive tasks for later execution via `handler.post` instead of updating queue that is being drained (that may lead to deadlock in core).

Reply:
thanks!